code/game/machinery/doors/firedoor.dm

```diff
@@ -44,29 +76,142 @@
 
-/obj/machinery/door/firedoor/proc/CalculateAffectingAreas()
-	remove_from_areas()
-	affecting_areas = get_adjacent_open_areas(src) | get_area(src)
-	for(var/I in affecting_areas)
-		var/area/A = I
-		LAZYADD(A.firedoors, src)
+///////////////////////////////////////////////////////////////////
+// Merger handling
 
-/obj/machinery/door/firedoor/closed
-	icon_state = "door_closed"
-	density = TRUE
+/obj/machinery/door/firedoor/proc/MergerAdding(obj/machinery/door/firedoor/us, datum/merger/new_merger)
+	SIGNAL_HANDLER
+	if(new_merger.id != merger_id)
+		return
+	RegisterSignal(new_merger, COMSIG_MERGER_REFRESH_COMPLETE, .proc/MergerRefreshComplete)
 
-//see also turf/AfterChange for adjacency shennanigans
+/obj/machinery/door/firedoor/proc/MergerRemoving(obj/machinery/door/firedoor/us, datum/merger/old_merger)
+	SIGNAL_HANDLER
+	if(old_merger.id != merger_id)
+		return
+	UnregisterSignal(old_merger, COMSIG_MERGER_REFRESH_COMPLETE)
 
+/// Handles the firelocks to register only one signal per group
+/obj/machinery/door/firedoor/proc/MergerRefreshComplete(datum/merger/merger, list/leaving_members, list/joining_members)
+
+///////////////////////////////////////////////////////////////////
+// End of Merger
+
+/**
+ * Calculates what areas we should worry about.
+ *
+ * This proc builds a list of areas we are in and areas we border
+ * and writes it to affecting_areas.
+ */
+/obj/machinery/door/firedoor/proc/CalculateAffectingAreas()
+	remove_from_areas()
+	affecting_areas = get_adjacent_open_areas(src) | get_area(src)
+	for(var/area/place in affecting_areas)
+		LAZYADD(place.firedoors, src)
+
+/**
+ * Removes us from any lists of areas in the affecting_areas list, then clears affecting_areas
+ *
+ * Undoes everything done in the CalculateAffectingAreas() proc, to clean up prior to deletion.
+ * Calls reset() first, in case any alarms need to be cleared first.
+ */
 /obj/machinery/door/firedoor/proc/remove_from_areas()
+	reset() //This handles some alert/alarm clearing
 	if(affecting_areas)
-		for(var/I in affecting_areas)
-			var/area/A = I
-			LAZYREMOVE(A.firedoors, src)
+		for(var/area/place in affecting_areas)
+			LAZYREMOVE(place.firedoors, src)
+			LAZYREMOVE(place.active_firelocks, src)
+			if(!LAZYLEN(place.active_firelocks)) //if we were the last firelock still active in this particular area
+				for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+					fire_panel.set_status()
+
+/obj/machinery/door/firedoor/should_atmos_process(datum/gas_mixture/air, exposed_temperature)
+	return world.time > detect_cooldown && (exposed_temperature > FIRE_MINIMUM_TEMPERATURE_TO_EXIST || exposed_temperature < BODYTEMP_COLD_DAMAGE_LIMIT) && !(obj_flags & EMAGGED) && !machine_stat
+
+/obj/machinery/door/firedoor/atmos_expose(datum/gas_mixture/air, exposed_temperature)
+	if(obj_flags & EMAGGED)
+		return
+	for(var/area/place in affecting_areas)
+		if(!place.fire_detect) //if any area is set to disable detection
+			return
+	if(alarm_type)
```

Comment:
Hey, it's now tomorrow + 10 days.
atmos_sensitive was an element I created to make dealing with code that wants to process while some atmos related check is true as easy as possible.
I don't think it fits this usecase, since you don't actually need to process. The environment is either inhospitable or it's not, right?

Reply:
Updated to not use atmos sensitive